### PR TITLE
feat(plugins): Baseline PF4J-backed plugins implementation

### PR DESCRIPTION
**Pull request by robzienert**
cc @link108 @ajordens

### CODE REVIEW

File: kork-plugins/src/main/kotlin/com/netflix/spinnaker/kork/plugins/ExtensionsInjector.kt [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    for (plugin in startedPlugins) {
+      log.debug("Registering extensions of the plugin '{}' as beans", plugin.pluginId)
+      extensionClassNames = pluginManager.getExtensionClassNames(plugin.pluginId)
+      for (extensionClassName in extensionClassNames) {
+        try {
+          log.debug("Register extension '{}' as bean", extensionClassName)
+          val extensionClass = plugin.pluginClassLoader.loadClass(extensionClassName)
+          registerExtension(extensionClass)
+        } catch (e: ClassNotFoundException) {
+          log.error(e.message, e)
```
Comment (dreynaud) [inline kork-plugins/src/main/kotlin/com/netflix/spinnaker/kork/plugins/ExtensionsInjector.kt]: minor: repeating `e.message` doesn't add a whole lot of context-schlontext. I'd go with something like

```suggestion
          log.error("Failed to register extension '{}' of plugin '{}'", extensionClassName, plugin.pluginId, e)
```

File: kork-plugins/src/test/kotlin/com/netflix/spinnaker/kork/plugins/ExtensionsInjectorTest.kt [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+import dev.minutest.junit.JUnit5Minutests
+import dev.minutest.rootContext
+import io.mockk.every
+import io.mockk.mockk
+import org.pf4j.PluginWrapper
+import strikt.api.expectThat
+import strikt.assertions.isA
+import strikt.assertions.isEqualTo
+
+class ExtensionsInjectorTest : JUnit5Minutests {
```
Comment (dreynaud) [inline kork-plugins/src/test/kotlin/com/netflix/spinnaker/kork/plugins/ExtensionsInjectorTest.kt]: read that as 5-Minutes-tests 🤔

File: kork-plugins/src/main/kotlin/com/netflix/spinnaker/config/PluginsConfigurationProperties.kt [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+ */
+@ConfigurationProperties("spinnaker.plugins")
+class PluginsConfigurationProperties {
+  /**
+   * Whether or not plugins are enabled for a particular service.
+   */
+  var enabled: Boolean = false
+
+  /**
+   * The root filepath to the directory containing all standard plugins.
```
[the remark this answers is not in the document]
Comment (robzienert) [reply kork-plugins/src/main/kotlin/com/netflix/spinnaker/config/PluginsConfigurationProperties.kt]: This is an artifact of a prior design where I had added support for multiple plugin root paths. That's not the case anymore so I'll update the docs.

File: kork-plugins/src/main/kotlin/com/netflix/spinnaker/kork/plugins/ExtensionsInjector.kt [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+ * enabled plugins into their correct places.
+ */
+class ExtensionsInjector(
+  private val pluginManager: PluginManager,
+  private var registry: BeanDefinitionRegistry
+) {
+
+  private val log by lazy { LoggerFactory.getLogger(javaClass) }
+
+  fun injectExtensions() {
```
[the remark this answers is not in the document]
Comment (robzienert) [reply kork-plugins/src/main/kotlin/com/netflix/spinnaker/kork/plugins/ExtensionsInjector.kt]: That's definitely true. I don't expect this codebase as it exists to be long-lived: My goal in this PR is to get a foundation in place that we can then divide/conquer iteratively as we need to.

**Comment by robzienert** [comment]
Flagged as `do-not-merge`. Netflix will be taking a different direction with plugin runtimes in Q4, pivoting to embracing remote processes for plugins.

We had a sync internally about plugins yesterday and decided PF4J is not what we'll be going forward with in Q4. After going over the POC, we concluded that while it's a good system, the problems of in-process plugins outweigh the benefits for an initial implementation. Although, if you feel strongly about having in-proc plugins like today, I do think PF4J is an excellent choice and we can certainly make it a plugin executor (but Netflix will not be prioritizing it as the first). The biggest problems we identified with in-proc strategies was failing to find a good experience around plugin observability, an uncompelling rollback story for Spinnaker operators (as opposed to plugin developers), and some non-trivial challenges around security.

Instead, we'll be looking at containers / functions as our initial plugin runtime (which hasn't been finalized yet): We feel that the tooling story around these technologies will be better for external developers than a more unfamiliar toolchain for JVM in-proc plugins. The approach we'll take is:

1. Define contracts for strictly-enforced extension points within Spinnaker (conceptually no deviation here). These will be in the form of protobuf `message` types, although what the shape of these messages look like, we're still not yet sure. Making the contracts just `message` gives us a little more latitude to implement different transports between plugin runtimes. Plugins will not be free-for-all functionality; if someone wants to extend Spinnaker in advanced ways, the existing Spring extension story still seems the best strategy - although we need to make it clearer how this can be done, with examples.
2. TCKs (technology compatibility kit) for both sides of the plugin (the extension provider and the extension plugin). One side of the TCK will ensure that a service's advertised interactions with plugins remains consistent. The second side of the TCK will allow plugin developers to assert their plugin functionality against this contract (beyond just the interface contract defined via protobuf).
3. Finally, once these prior steps are complete, we'll implement our first runtime, which we'll likely be targeting containers.

We're thinking this work will be done against two different projects: Execution Stages, and an operator's plugin view / plugin developer playground. The latter is still in flux as to what its final capabilities will entail.

This is a bit of a departure from what we were initially planning to do for Q4, but doesn't fundamentally change the key tasks and challenges. I'll be working on clarifying various aspects of what I laid out above next week, but wanted to drop something (not) quick into the channel now in case someone sees this and wants to drop some thoughts.

**Comment by robzienert** [comment]
Unmarked as `do not merge`. After this last week, we were able to identify some use cases for this quarter that would best fit with in-process extensions (SpEL expressions and Micrometer backends are two obvious choices). Further, we will be using PF4J to implement and provide different plugin runtimes as well (in-process already done, will also use it for grpc, functions, etc). We'll still be moving forward with a remote runtime as well.

I'll be writing up a proposal this week to outline the general direction we're proposing, and how it fits in with the existing plugin work.

**Pull request merged**